etl/src/main/java/org/pih/hivmigration/etl/sql/ExamLabResultsMigrator.groovy

```diff
@@ -243,6 +240,61 @@ class ExamLabResultsMigrator extends ObsMigrator {
         ''')
 
         migrate_tmp_obs()
+
+        //
+        // -------- Hemoglobin (and some hematocrit)
+        //
+
+        create_tmp_obs_table()
+
+        executeMysql("Migrate hemoglobin values", '''
+            INSERT INTO tmp_obs (source_encounter_id, concept_uuid, value_numeric)
+            SELECT
+                he.source_encounter_id,
+                concept_uuid_from_mapping('PIH', 'HEMOGLOBIN'),
+                extract_number(result)
+            FROM hivmigration_exam_lab_results helr
+            JOIN hivmigration_encounters he on helr.source_encounter_id = he.source_encounter_id
+            JOIN encounter e on he.encounter_id = e.encounter_id
+            WHERE lab_test = 'hemoglobin'
+                AND (result REGEXP '^[[:blank:]]*[0-9\\.]+[[:blank:]]*gr?.dll?$'  -- e.g. '9.8 g/dl' or '9.8gr/dl'
+                     OR is_number(result))  -- e.g. 9.8
+            ;
+        ''')
+
+        executeMysql("Migrate hematocrit values incorrectly marked as hemoglobin", '''
+            INSERT INTO tmp_obs (source_encounter_id, concept_uuid, value_numeric)
+            SELECT
+                he.source_encounter_id,
+                concept_uuid_from_mapping('PIH', 'HEMATOCRIT'),
+                extract_number(result)
+            FROM hivmigration_exam_lab_results helr
+            JOIN hivmigration_encounters he on helr.source_encounter_id = he.source_encounter_id
+            JOIN encounter e on he.encounter_id = e.encounter_id
+            WHERE lab_test = 'hemoglobin'
+                AND result REGEXP '^(Hte?|HTE?)?[[:blank:]]?[0-9\\.]+%$'  -- e.g. '33%' or 'Ht 33.0%'
+            ;
+        ''')
+
+        executeMysql("Log warning about invalid hemoglobin values", '''
+            INSERT INTO hivmigration_data_warnings (openmrs_patient_id, openmrs_encounter_id, field_name, field_value, warning_type)
+            SELECT
+                hp.person_id,
+                he.encounter_id,
+                'HIV_EXAM_LAB_RESULTS hemoglobin',
+                helr.result,
+                'Invalid hemoglobin test value. Not migrated.'
+            FROM hivmigration_exam_lab_results helr
+            JOIN hivmigration_encounters he ON helr.source_encounter_id = he.source_encounter_id
+            JOIN hivmigration_patients hp ON he.source_patient_id = hp.source_patient_id
+            WHERE lab_test = 'hemoglobin'
+                  AND result NOT regexp '^[[:blank:]]*[0-9\\.]+[[:blank:]]*gr?.dll?$'
+                  AND result NOT regexp '^(Hte?|HTE?)?[[:blank:]]?[0-9\\.]+%$'
+                  AND NOT is_number(result)
+                  AND NOT is_number(REPLACE(result, ',', '.'));
+          ''')
```

Comment:
I like @mseaton 's idea from the other PR of not doing an inverse query here, but just seeing what rows didn't make it into tmp_obs.

Reply:
Unfortunately since there's no unique identifier for those rows that goes into tmp_obs, it's not so easy to do that query. I've got an improvement in the works. Coming tomorrow morning.

Reply:
https://github.com/PIH/hivmigration/pull/77